### PR TITLE
Add shock intial condition in functional tests

### PR DESCRIPTION
**Pull request by nicolasaunai**
add shock functional test.

Add Tagger for automatic refinement
MultiphysicsIntegrator uses the Tagger, which is abstract and actually depends on which model is being solved at current level.
There are therefore concrete taggers, per model. Each concrete Tagger can then have multiple tagging strategies. Only 1 is now coded. (Other attempts may come later, and also may be taken from user supplied python tagging methods.)

Bug fixes

    first bug fix concerns EM_old and Jold in the hybridMessenger. These quantities are supposed to store the current state of those model variables at time n before the level advances to time n+1. They are then used by next finer level to get ghosts at subcycle times which are per definition between coarse steps n and n+1. before this PR, the "old" quantities had their patchdata timestamp fixed at the creation of the level but never updated afterwards. The current PR thus sets their timestamps in prepareStep, i.e. before every advance().

    fix bug concerning regriding: Regriding means taking a level out of the hierarchy and replacing it by a new one. This action invalidates all the schedules stored in the RefinerPool that deal with the level being replaced. This PR now re-create schedules for the level that's being regrided and the next finer as well. This is important because SAMRAI typically will do things in the following order:

    for a hierarchy with 2 levels:

    tag level 1

    create level 2

    regrid level 1

thus at the time L2 is created, its schedules are made with the L1 taht is about to be taken out of the hierarchy. This PR now re-makes these schedules when regriding L1.

    the PR also adds particles to the maxwellian initializer test as well as prints, in an attempt to know more about latest (random?) failures observed of this tests...

tagging :
The concrete tagging strategy that is implemented tags a cell (==1) when the local 3 points average of the magnetic field differs from the current value from more than 10%. It's a first attempt that seems to give consistent results in tha tdtagged functional test but will need further attention in upcoming PRs.

Notes :

    The concrete tagging strategy ignores what samrai stuff is, and probably could be moved to a core lib section. That should be considered later as this matter possibly also concern other components.

### CODE REVIEW

**Comment by nicolasaunai** [comment]
LGTM "checkout issue" : https://github.com/github/codeql/issues/4809

File: pyphare/pyphare/core/gridlayout.py [hunk 1]
```diff
@@ -11,20 +11,22 @@
...
+        "Fx":"primal","Fy":"primal","Fz":"primal",
         'Vz':'primal','P':'primal'
     },
     'y' : {
         'Bx':'dual', 'By':'primal', 'Bz':'dual',
         'Ex':'primal', 'Ey':'dual', 'Ez':'primal',
         'Jx':'primal', 'Jy':'dual', 'Jz':'primal',
         'rho':'primal', 'Vx':'primal','Vy':'primal',
-        'Vz':'primal','P':'primal'
+        'Vz':'primal','P':'primal', "Fx":"primal","Fy":"primal","Fz":"primal"
```
Comment (PhilipDeegan) [inline pyphare/pyphare/core/gridlayout.py:22]: the odd one out

File: pyphare/pyphare/pharesee/hierarchy.py [hunk 2]
```diff
@@ -179,6 +179,47 @@ def level_range(self):
...
+    assert hierarchy.dim == 1
+
+    x_ = pdata.x
+    v_ = pdata.dataset
+
+    if ilvl == hierarchy.finest_level():
+        return x_,v_
+
+    qtyname = pdata.field_name
+    inner = x_ != x_ #lgtm [py/comparison-of-identical-expressions]
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharesee/hierarchy.py:203]: whats' this doing?

File: pyphare/pyphare/pharesee/hierarchy.py [hunk 3]
```diff
@@ -179,6 +179,47 @@ def level_range(self):
...
+
+    # iteratively fill the mask with true where current patch coordinates
+    # are within limits of the next refined level patchdatas
+    for ipatch, finer_patch in enumerate(hierarchy.patch_levels[ilvl+1].patches):
+        pdat = finer_patch.patch_datas[qtyname]
+        xmin,xmax = [pdat.x[ix] for ix in (4,-4)]
+        inner  = inner | ((x_ > xmin) & (x_ < xmax))
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharesee/hierarchy.py:210]: not so obvious

File: pyphare/pyphare/simulator/simulator.py [hunk 4]
```diff
@@ -96,7 +96,7 @@ def run(self):
         print("mean advance time = {}".format(np.mean(perf)))
         print("total advance time = {}".format(np.sum(perf)))
 
-        print(perf)
+        #print(perf)
```
Comment (PhilipDeegan) [inline pyphare/pyphare/simulator/simulator.py:99]: rm

File: src/amr/tagging/default_hybrid_tagger_strategy.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+namespace PHARE::amr
+{
+template<typename HybridModel>
+class DefaultHybridTaggerStrategy : public HybridTaggerStrategy<HybridModel>
+{
+    using gridlayout_type           = typename HybridModel::gridlayout_type;
+    static auto constexpr dimension = HybridModel::dimension;
+
+public:
+    void tag(HybridModel& model, gridlayout_type const& layout, int* tags) const override;
```
Comment (PhilipDeegan) [inline src/amr/tagging/default_hybrid_tagger_strategy.h:18]: is this interacting with SAMRAI somehow that it has to be a raw "int"?

File: src/amr/tagging/default_hybrid_tagger_strategy.h [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    double threshold = 0.05;
+    int idx          = 4;
+    double epsilon   = 0.1;
+
+
+    if constexpr (dimension == 1)
+    {
+        for (auto iCell = 0u, ix = start; iCell <= endCell; ++ix, ++iCell)
+        {
+            auto Byavgm2 = 0.2 * (By(ix - 4) + By(ix - 3) + By(ix - 2) + By(ix - 1) + By(ix));
```
Comment (PhilipDeegan) [inline src/amr/tagging/default_hybrid_tagger_strategy.h:46]: is there some documentation to go along with this as a reference?

it's pretty complicated

File: src/solver/multiphysics_integrator.h [hunk 7]
```diff
@@ -367,7 +391,7 @@ namespace solver
             // so dividing the mesh size by ratio means dt
             // needs to be divided by ratio^2.
             // we multiply that by a constant < 1 for safety.
-            return coarseDt / (ratio.max() * ratio.max()) * 0.4;
+            return coarseDt / (ratio.max() * ratio.max()) /** 0.4*/;
```
Comment (PhilipDeegan) [inline src/solver/multiphysics_integrator.h:394]: TODO?

File: src/solver/multiphysics_integrator.h [hunk 8]
```diff
@@ -525,6 +550,21 @@ namespace solver
...
+            bool hasTagger = true;
+
+            for (auto iLevel = coarsestLevel; iLevel <= finestLevel; ++iLevel)
+            {
+                if (levelDescriptors_[iLevel].taggerIndex != LevelDescriptor::NOT_SET)
+                {
+                    return hasTagger;
+                }
+            }
+            return !hasTagger;
```
Comment (PhilipDeegan) [inline src/solver/multiphysics_integrator.h:564]: "hasTagger" seems a bit pointless

File: src/amr/tagging/default_hybrid_tagger_strategy.h [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            auto Navg   = 0.2 * (N(ix - 2) + N(ix - 1) + N(ix) + N(ix + 1) + N(ix + 2));
+
+            auto criter_by = std::abs(Byavgp1 - Byavg) / (1 + std::abs(Byavg));
+            auto criter_bz = std::abs(Bzavgp1 - Bzavg) / (1 + std::abs(Bzavg));
+            auto criter_b  = std::sqrt(criter_by * criter_by + criter_bz * criter_bz);
+            auto criter_n  = std::abs(Navgp1 - Navg) / (1 + std::abs(Navg));
+
+            auto criter = std::max(criter_b, criter_n);
+
+            if (criter > threshold)
```
Comment (PhilipDeegan) [inline src/amr/tagging/default_hybrid_tagger_strategy.h:68]:
```
 tags[iCell] = criter > threshold;
```

File: tests/amr/tagging/test_tagging.cpp [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        model.state.electromag.E.setBuffer("EM_E_y", &Ey);
+        model.state.electromag.E.setBuffer("EM_E_z", &Ez);
+        model.state.electromag.B.setBuffer("EM_B_x", &Bx);
+        model.state.electromag.B.setBuffer("EM_B_y", &By);
+        model.state.electromag.B.setBuffer("EM_B_z", &Bz);
+        model.state.electromag.initialize(layout);
+    }
+};
+
+/*
```
Comment (PhilipDeegan) [inline tests/amr/tagging/test_tagging.cpp:178]: TODO?

File: tests/functional/alfven_wave/alfven_wave1d.py [hunk 11]
```diff
@@ -165,9 +163,9 @@ def main():
     simulator.run()
 
 
-    if cpp.mpi_rank() == 0:
-        b = hierarchy_from(h5_filename="phare_outputs/EM_B.h5")
-        plot(b)
-
+    #if cpp.mpi_rank() == 0:
```
Comment (PhilipDeegan) [inline tests/functional/alfven_wave/alfven_wave1d.py:166]: unnecessary, or was there an issue?

File: src/amr/tagging/default_hybrid_tagger_strategy.h (hunk 5 above)
Comment (nicolasaunai) [reply src/amr/tagging/default_hybrid_tagger_strategy.h:18]: yes

File: tests/functional/alfven_wave/alfven_wave1d.py (hunk 11 above)
Comment (nicolasaunai) [reply tests/functional/alfven_wave/alfven_wave1d.py:166]: tbd

**Review by PhilipDeegan — APPROVED** [review]

**Pull request merged**